client/src/visualiser-src/graph-visualiser/data-structure/GraphicalGraph.ts

```diff
@@ -0,0 +1,196 @@
+    // in the visualiser canvas.
+    this.loadGraph();
+    return new GraphAnimationProducer();
+  }
+
+  /**
+   * Inserts the edge from `from` to `to`.
+   * Assumes that `from` and `to` are valid vertex numbers.
+   * TODO: currently unanimated. Does this need an animation?
+   * @param from
+   * @param to
+   * @returns
+   */
+  addEdge(from: number, to: number): AnimationProducer {
+    this.edges = [
+      ...this.edges,
+      { source: String(from), target: String(to), weight: 5, isBidirectional: false },
+    ];
+
+    // Reload the graph to sync the change in `this.edges` with what's shown
+    // in the visualiser canvas.
+    this.loadGraph();
+    return new GraphAnimationProducer();
+  }
+
+  /**
+   * Produces a sequence of animations for visualising DFS.
+   * @remarks
+   * This method wraps around the recursive helper, `dfsRecurse`.
+   * @param src The vertex to start DFS from.
+   * @returns Animation producer with the steps necessary for visualising DFS.
+   */
+  public dfs(src: number): AnimationProducer {
+    const producer = new GraphDfsAnimationProducer();
+    producer.renderDfsCode();
+
+    const visited = new Set<number>();
+    this.dfsRecurse(producer, src, visited);
+
+    return producer;
+  }
+
+  /**
+   * Tells the animation producer to produce the sequence of vertex and edge
+   * highlighting necessary to animate a DFS traversal of the current graph.
+   * @remarks
+   * Assumes that `visited` does not contain `src`.
+   * @param animationProducer
+   * @param src
+   * @param visited A set of all vertices that have been visited through DFS.
+   */
+  private dfsRecurse(
+    animationProducer: GraphDfsAnimationProducer,
+    src: number,
+    visited: Set<number>
+  ): void {
+    // Mark the current vertex as visited.
+    visited.add(src);
+    const vertexElem = Vertex.getDomReference(src);
+    animationProducer.doAnimationAndHighlight(2, animationProducer.highlightVertex, vertexElem);
+
+    // For each unvisited neighbour, highlight the edge to that neighbour and
+    // launch DFS on them.
+    for (let neighbour = 0; neighbour < this.vertices.length; neighbour += 1) {
+      if (this.isAdjacent(src, neighbour) && !visited.has(neighbour)) {
+        animationProducer.doAnimationAndHighlight(
+          4,
+          animationProducer.highlightEdge,
+          Edge.getDomReference(src, neighbour)
+        );
+        this.dfsRecurse(animationProducer, neighbour, visited);
+      }
+    }
+  }
+
+  /**
+   * Determines whether the edge from `v` to `w` exists.
+   * @param v Source vertex.
+   * @param w Destination vertex.
+   * @returns existence of edge v-w.
+   */
+  private isAdjacent(v: number, w: number): boolean {
+    return this.edges.some(
+      (edge) =>
+        (edge.source === String(v) && edge.target === String(w)) ||
+        (edge.source === String(w) && edge.target === String(v) && edge.isBidirectional)
+    );
+  }
+
+  public generate(): void {
```

Comment:
Done, but went with a different algorithm since I found that randomly generating nodes and edges produced pretty ugly graphs.

I went with 'random walk' (https://en.wikipedia.org/wiki/Random_walk) algorithm which was about as simple.